datafusion/core/src/physical_optimizer/sort_enforcement.rs

```diff
@@ -102,44 +128,189 @@ impl TreeNodeRewritable for PlanWithCorrespondingSort {
-                        // However, we want to propagate them above anyway.
-                        for (maintains, element) in flags.into_iter().zip(onwards.iter())
-                        {
-                            if (maintains || is_sort(&item.plan)) && !element.is_empty() {
-                                return element.clone();
-                            }
+                .map(|item| item.plan.clone())
+                .collect();
+            let coalesce_onwards = children_requirements
+                .into_iter()
+                .enumerate()
+                .map(|(idx, item)| {
+                    // Leaves of the `coalesce_onwards` tree are `CoalescePartitionsExec`
+                    // operators. This tree collects all the intermediate executors that
+                    // maintain a single partition. If we just saw a `CoalescePartitionsExec`
+                    // operator, we reset the tree and start accumulating.
+                    let plan = item.plan;
+                    if plan.as_any().is::<CoalescePartitionsExec>() {
+                        Some(ExecTree {
+                            idx,
+                            plan,
+                            children: vec![],
+                        })
+                    } else if plan.children().is_empty() {
+                        // Plan has no children, there is nothing to propagate.
+                        None
+                    } else {
+                        let children = item
+                            .coalesce_onwards
+                            .into_iter()
+                            .flatten()
+                            .filter(|item| {
+                                // Only consider operators that don't require a
+                                // single partition.
+                                !matches!(
+                                    plan.required_input_distribution()[item.idx],
+                                    Distribution::SinglePartition
+                                )
+                            })
+                            .collect::<Vec<_>>();
+                        if children.is_empty() {
+                            None
+                        } else {
+                            Some(ExecTree {
+                                idx,
+                                plan,
+                                children,
+                            })
                         }
                     }
-                    vec![]
                 })
-                .collect::<Vec<_>>();
+                .collect();
             let plan = with_new_children_if_necessary(self.plan, children_plans)?;
-            Ok(PlanWithCorrespondingSort { plan, sort_onwards })
+            Ok(PlanWithCorrespondingCoalescePartitions {
+                plan,
+                coalesce_onwards,
+            })
         }
     }
 }
 
```

Comment:
Same as above, suggest to move the construction of `coalesce_onwards` to a `init` method of `PlanWithCorrespondingCoalescePartitions` and keep the `map_children()` simple

Reply:
Moved these to init method.